RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc

```diff
@@ -0,0 +1,594 @@
+  }
+
+  // compute "max z" for first layer in endcap (should we restrict to the outermost ring?)
+  for (auto im = phase1PixelTopology::layerStart[4]; im < phase1PixelTopology::layerStart[5]; ++im) {
+    auto const& g = m_detParamsGPU[im];
+    aveGeom.endCapZ[0] = std::max(aveGeom.endCapZ[0], g.frame.z());
+  }
+  for (auto im = phase1PixelTopology::layerStart[7]; im < phase1PixelTopology::layerStart[8]; ++im) {
+    auto const& g = m_detParamsGPU[im];
+    aveGeom.endCapZ[1] = std::min(aveGeom.endCapZ[1], g.frame.z());
+  }
+  // correct for outer ring being closer
+  aveGeom.endCapZ[0] -= 1.5f;
+  aveGeom.endCapZ[1] += 1.5f;
+
+  /*
+  for (int jl=0, nl=phase1PixelTopology::numberOfLaddersInBarrel; jl<nl; ++jl) {
+    std::cout << jl<<':'<<aveGeom.ladderR[jl] << '/'<< std::sqrt(aveGeom.ladderX[jl]*aveGeom.ladderX[jl]+aveGeom.ladderY[jl]*aveGeom.ladderY[jl]) 
+                   <<','<<aveGeom.ladderZ[jl]<<','<<aveGeom.ladderMinZ[jl]<<','<<aveGeom.ladderMaxZ[jl]<< ' ';
+  } std::cout<< std::endl;
+  std::cout << aveGeom.endCapZ[0] << ' ' << aveGeom.endCapZ[1] << std::endl;
+  */
+
+  // fill Layer and ladders geometry
+  memcpy(m_layerGeometry.layerStart, phase1PixelTopology::layerStart, sizeof(phase1PixelTopology::layerStart));
+  memcpy(m_layerGeometry.layer, phase1PixelTopology::layer.data(), phase1PixelTopology::layer.size());
+}
+
+PixelCPEFast::GPUData::~GPUData() {
+  if (d_paramsOnGPU != nullptr) {
+    cudaFree((void*)h_paramsOnGPU.m_commonParams);
+    cudaFree((void*)h_paramsOnGPU.m_detParams);
+    cudaFree((void*)h_paramsOnGPU.m_averageGeometry);
+    cudaFree((void*)h_paramsOnGPU.m_layerGeometry);
+    cudaFree(d_paramsOnGPU);
+  }
+}
+
+std::unique_ptr<PixelCPEBase::ClusterParam> PixelCPEFast::createClusterParam(const SiPixelCluster& cl) const {
+  return std::make_unique<ClusterParamGeneric>(cl);
+}
+
+void PixelCPEFast::errorFromTemplates(DetParam const& theDetParam,
+                                      ClusterParamGeneric& theClusterParam,
+                                      float qclus) const {
+  float locBz = theDetParam.bz;
+  float locBx = theDetParam.bx;
+  //cout << "PixelCPEFast::localPosition(...) : locBz = " << locBz << endl;
+
+  theClusterParam.pixmx = std::numeric_limits<int>::max();  // max pixel charge for truncation of 2-D cluster
+
+  theClusterParam.sigmay = -999.9;  // CPE Generic y-error for multi-pixel cluster
+  theClusterParam.sigmax = -999.9;  // CPE Generic x-error for multi-pixel cluster
+  theClusterParam.sy1 = -999.9;     // CPE Generic y-error for single single-pixel
+  theClusterParam.sy2 = -999.9;     // CPE Generic y-error for single double-pixel cluster
+  theClusterParam.sx1 = -999.9;     // CPE Generic x-error for single single-pixel cluster
+  theClusterParam.sx2 = -999.9;     // CPE Generic x-error for single double-pixel cluster
+
+  float dummy;
+
+  SiPixelGenError gtempl(thePixelGenError_);
+  int gtemplID_ = theDetParam.detTemplateId;
```

Comment:
```suggestion
  int gtemplID = theDetParam.detTemplateId;
```
trailing underscore reserved for data members

Reply:
Done in https://github.com/cms-patatrack/cmssw/pull/593/commits/b7c587960302b89e417aad3df6bf00294bb888e2 .